compiler/src/iree/compiler/Codegen/WGSL/WGSLReplacePushConstants.cpp

```diff
@@ -0,0 +1,137 @@
+  if (opType.isIndex()) {
+    auto indexCastOp =
+        builder.create<arith::IndexCastOp>(op.getLoc(), opType, extractOp);
+    op.replaceAllUsesWith(indexCastOp.getResult());
+  } else {
+    // TODO(scotttodd): truncate or zero-extend types with different bitwidths?
+    auto bitcastOp =
+        builder.create<arith::BitcastOp>(op.getLoc(), op.getType(), extractOp);
+    op.replaceAllUsesWith(bitcastOp.getResult());
+  }
+
+  op.erase();
+}
+
+class WGSLReplacePushConstantsPass
+    : public WGSLReplacePushConstantsBase<WGSLReplacePushConstantsPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<mlir::arith::ArithmeticDialect, mlir::func::FuncDialect,
+                    mlir::tensor::TensorDialect, IREE::Flow::FlowDialect,
+                    IREE::HAL::HALDialect>();
+  }
+
+  void runOnOperation() override {
+    auto parentOp = getOperation();
+    auto loc = parentOp.getLoc();
+    auto constantLoadOps = llvm::to_vector<4>(
+        parentOp.getOps<IREE::HAL::InterfaceConstantLoadOp>());
+    if (constantLoadOps.empty()) return;
+
+    OpBuilder builder(parentOp);
+    builder.setInsertionPointToStart(&parentOp.getBlocks().front());
+
+    // Group all push constants into a single `hal.interface.binding.subspan`
+    // and load from it once using `flow.dispatch.tensor.load`, then extract
+    // individual push constants with `tensor.extract`.
+
+    // Find the range of push constant indices (0 to some maximum).
+    uint64_t maxConstantIndex = 0;
+    // Inspect the alignment values. These are just hints, so if all are equal
+    // then use the value, otherwise drop the alignment hint.
+    SmallVector<uint64_t> alignmentValues;
+    bool missingAlignmentValue = false;
+    for (auto constantLoadOp : constantLoadOps) {
+      maxConstantIndex =
+          std::max(constantLoadOp.getIndex().getZExtValue(), maxConstantIndex);
+
+      auto alignmentAttr = constantLoadOp.getAlignmentAttr();
+      if (alignmentAttr) {
+        uint64_t alignmentValue = alignmentAttr.getValue().getZExtValue();
+        alignmentValues.push_back(alignmentValue);
+      } else {
+        missingAlignmentValue = true;
+      }
+    }
+    auto maxConstantValue =
+        builder.createOrFold<arith::ConstantIndexOp>(loc, maxConstantIndex);
+    mlir::IntegerAttr alignmentAttr = nullptr;
+    // TODO(scotttodd): try llvm::all_equal with attrs directly
+    if (!missingAlignmentValue && llvm::all_equal(alignmentValues)) {
+      alignmentAttr = constantLoadOps[0].getAlignmentAttr();
+    }
+
+    // hal.interface.binding.subspan -> !flow.dispatch.tensor<readonly:i32>
+    //   * Group all push constants into a single tensor<Nxi32>
+    //   * If individual data types differ, they'll be bitcast when extracted
+    auto dispatchTensorType = IREE::Flow::DispatchTensorType::get(
+        IREE::Flow::TensorAccess::ReadOnly, {}, builder.getI32Type());
+    SmallVector<Value> dynamicDims;
```

Comment:
Drop `dynamicDims` and just use `{}` everywhere. Less confusing?

Reply:
I tried that, but I get compiler errors like
```
[build] /usr/local/google/home/scotttodd/code/iree/compiler/src/iree/compiler/Codegen/WGSL/WGSLReplacePushConstants.cpp:110:30: error: no matching member function for call to 'create'
[build]     auto subspanOp = builder.create<IREE::HAL::InterfaceBindingSubspanOp>(
[build]                      ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[build] /usr/local/google/home/scotttodd/code/iree/third_party/llvm-project/mlir/include/mlir/IR/Builders.h:456:8: note: candidate template ignored: substitution failure [with OpTy = mlir::iree_compiler::IREE::HAL::InterfaceBindingSubspanOp]: deduced incomplete pack <mlir::iree_compiler::IREE::Flow::DispatchTensorType &, llvm::APInt, llvm::APInt, mlir::iree_compiler::IREE::HAL::DescriptorType, mlir::arith::ConstantIndexOp &, (no value), mlir::IntegerAttr &> for template parameter 'Args'
[build]   OpTy create(Location location, Args &&...args) {
[build]        ^
[build] 1 error generated.
```

Reply:
Ah Windows. You probably need to do `ArrayRef<Value>{}` instead of just `{}`